### PR TITLE
Pipeline Portability Charter

### PR DESCRIPTION
**Pull request by TimothyTickle**
Summary of major points of discussion:
- Language was updated to better communicate ideas but core concepts were not changed.
- Definitions were added for clarity.
- Formatting

**February 26:** Last call for oversight reviews

### CODE REVIEW

File: charters/PORTABILITY/charter.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Pipeline Portability Service
```
Comment (brianraymor) [inline charters/PORTABILITY/charter.md]: It seems like the directory should be named _PipelinePortabilityService_ or _PIpelinePortability_ rather than _PORTABILITY_?

File: charters/PORTABILITY/charter.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+## Description
+
+The portability service enables the confirmation that pipelines developed for the HCA DCP can be successfully ran outside of the HCA DCP.
```
Comment (brianraymor) [inline charters/PORTABILITY/charter.md]: Change all instances of _portability service_ to _Portability Service_?

Editorial - what about  _The portability services confirms (or validates) that pipelines developed for the HCA DCP can successfully execute outside the HCA DCP_ - a bit more context would be helpful. I expect some people to quickly browse the **Description** for the elevator pitch. Why is this _goodness_? Where will the pipelines execute?

File: charters/PORTABILITY/charter.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+[Kylee Degatano](mailto:kdegatano@broadinstitute.org)
+
+### Technical Lead
+[Marcus Kinsella](mailto:mkinsella@chanzuckerberg.com)
+
+## Communication
+
+### Slack Channels
+
+__HumanCellAtlas/portability:__ To discuss portability and access of pipelines to infrastructure outside Lira.
```
Comment (brianraymor) [inline charters/PORTABILITY/charter.md]: I love that all your charters include descriptions for the repo(s) - could _Lira_ be defined or replaced with a general term?

File: charters/PORTABILITY/charter.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+## Objectives
+
+The portability service provides the ability to run HCA DCP pipelines 
+
+## In-scope
+
+* Programmatic authorized access to submit pipelines that will be ran to confirm they execute to completion on different computing infrastructure (execution environments)
+* Programmatic authorized access to extend the backend of the portability service to include additional execution environments
+* A UI to describe the current state of submitted pipelines and execution environments
+* Documentation to describe how to submit a backend or pipeline the system
```
Comment (brianraymor) [inline charters/PORTABILITY/charter.md]: Is there a missing 'to' - as in _how to submit a backend or pipeline **to** the system_ ?

Were you planning to define _pipeline_?

File: charters/PORTABILITY/charter.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+* Documentation on portability best practices
+
+### Scientific "guardrails"
+
+Portability service will not comment on scientific accuracy of pipelines, with this constraint it will not need guardrails
```
Comment (brianraymor) [inline charters/PORTABILITY/charter.md]: The use of _comment_ struck me as odd for a service that's not Siri. How about:

- will not address the accuracy
- will not report on the accuracy
- will not assess the accuracy
- will not validate the accuracy

File: charters/PORTABILITY/charter.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+* Updating backends so they support pipeline execution
+* Updating pipelines so they are portable to environments that do not support workflow language standards
+* Establishing standard data sets for portability (they are provided per pipeline)
+* Adding 3rd party backends
+
+## Milestones and Deliverables
+
+__2018:__ Authenticated access is established for an endpoint to submit pipelines.   
+__2018:__ Basic UI is established for pipeline and backend status.   
+__2019:__ Portability service is leveraged in the CI/CD of all pipelines ran by Lira.
```
Comment (brianraymor) [inline charters/PORTABILITY/charter.md]: What is _Lira_?

File: charters/PORTABILITY/charter.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+[Kylee Degatano](mailto:kdegatano@broadinstitute.org)
+
+### Technical Lead
+[Marcus Kinsella](mailto:mkinsella@chanzuckerberg.com)
+
+## Communication
+
+### Slack Channels
+
+[HumanCellAtlas/portability](https://humancellatlas.slack.com/messages/portability): To discuss portability and access of pipelines to infrastructure outside Lira.
```
Comment (brianraymor) [inline charters/PORTABILITY/charter.md]: What is _Lira_?

File: charters/PORTABILITY/charter.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+### Scientific "guardrails"
+
+Portability service will not comment on scientific accuracy of pipelines, with this constraint it will not need guardrails
+
+## Out-of-scope
+
+* Portability service will not comment on scientific accuracy of pipelines
+* Educational training or user assistance on how to create pipelines
+* Updating backends so they support pipeline execution
+* Updating pipelines so they are portable to environments that do not support workflow language standards
```
Comment (brianraymor) [inline charters/PORTABILITY/charter.md]: It's odd that _workflow language standards_ are not referenced in the in-scope section. Should this be called out?

File: charters/PORTABILITY/charter.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
+# Pipeline Portability Service
+
+## Description
+
+The portability service enables the confirmation that pipelines developed for the HCA DCP can be successfully ran outside of the HCA DCP.
+
+## Objectives
+
+The portability service provides the ability to run HCA DCP pipelines
```
Comment (ambrosejcarr) [inline charters/PORTABILITY/charter.md]: I found this hard to parse. Where do they run? Why? This description seems like it would better fit the Data Processing Service.

File: charters/PORTABILITY/charter.md (hunk 2 above)
Comment (ambrosejcarr) [reply charters/PORTABILITY/charter.md]: `s/ran/run` for future tense assertions about running pipelines, I think.

File: charters/PORTABILITY/charter.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+## Objectives
+
+The portability service provides the ability to run HCA DCP pipelines 
+
+## In-scope
+
+* Programmatic authorized access to submit pipelines that will be ran to confirm they execute to completion on different computing infrastructure (execution environments)
+* Programmatic authorized access to extend the backend of the portability service to include additional execution environments
+* A UI to describe the current state of submitted pipelines and execution environments
```
Comment (ambrosejcarr) [inline charters/PORTABILITY/charter.md]: If the UI is descriptive, is this just a website? Or are you proposing something more interactive?

File: charters/PORTABILITY/charter.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+* Documentation to describe how to submit a backend or pipeline the system
+* Documentation on portability best practices
+
+### Scientific "guardrails"
+
+Portability service will not comment on scientific accuracy of pipelines, with this constraint it will not need guardrails
+
+## Out-of-scope
+
+* Portability service will not comment on scientific accuracy of pipelines
```
Comment (ambrosejcarr) [inline charters/PORTABILITY/charter.md]: Redundant with above guardrails section -- I think this can be removed.

File: charters/PORTABILITY/charter.md (hunk 8 above)
Comment (ambrosejcarr) [inline charters/PORTABILITY/charter.md]: Should "workflow language standards" be defined somewhere?

File: charters/PORTABILITY/charter.md [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+* Educational training or user assistance on how to create pipelines
+* Updating backends so they support pipeline execution
+* Updating pipelines so they are portable to environments that do not support workflow language standards
+* Establishing standard data sets for portability (they are provided per pipeline)
+* Adding 3rd party backends
```
Comment (ambrosejcarr) [inline charters/PORTABILITY/charter.md]: How does this statement interact with https://github.com/HumanCellAtlas/dcp-community/pull/32/files#diff-304524626202982f9acc745b641f4961R14 ? 

What is a 3rd party backend and how is it different from the in-scope backends?

File: charters/PORTABILITY/charter.md [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+## Milestones and Deliverables
+
+__2018:__ Authenticated access is established for an endpoint to submit pipelines.   
+__2018:__ Basic UI is established for pipeline and backend status.   
+__2019:__ Portability service is leveraged in the CI/CD of all pipelines ran by Lira.   
+__2019:__ All backends supported by the HCA project are attached.
```
Comment (ambrosejcarr) [inline charters/PORTABILITY/charter.md]: Are these enumerated somewhere? Can this be linked to?

File: charters/PORTABILITY/charter.md (hunk 1 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: Updated, will be in next push.

File: charters/PORTABILITY/charter.md (hunk 2 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: Updated with more context on what principle we are trying to enable and that the system is focused on external execution and not internal testing.

using executed instead of run but thank you for the suggestion.

File: charters/PORTABILITY/charter.md (hunk 3 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: adding a definitions section, will include Lira

File: charters/PORTABILITY/charter.md (hunk 5 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: Lol, ok fair enough. I changed to using some of your power words (assess and report), see in the upcoming push.

File: charters/PORTABILITY/charter.md (hunk 6 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: Will add to definitions.

File: charters/PORTABILITY/charter.md (hunk 7 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: Will add to definitions.

File: charters/PORTABILITY/charter.md (hunk 8 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: This service really is not about best language practices, just reproducibility (can the pipelines execute successfully to completion). Certainly, as a by-product, lessons learned from the service have and will continue to contribute to moving forward best practices in workflow language features or inform on the features we use/recommend.

File: charters/PORTABILITY/charter.md (hunk 9 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: I agree, this text makes the service feel like a custom pipeline service (which it is not). Let me update with more context.

File: charters/PORTABILITY/charter.md (hunk 10 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: The first draft of the UI would simply be state. I would imagine the most interactive thing it would have is a login. This is however close to implementation detail and I want to be careful not to go into implementation. The kinds of questions it would answer are 1. what are your pipelines and did they pass or fail and in what environment 2. what are your backends, how many pipelines passed or failed. Also maybe access to your run logs. We would have to work with users for anything more. I could see it being helpful for a backend to see the kinds of errors pipelines they have in their system but that would be contingent on if the pipeline developers would be ok with them seeing that. I would imagine yes but seems like an opt-in kinda thing. Either way, this would be something we iterate forward on after the basic UI stated here.

File: charters/PORTABILITY/charter.md (hunk 11 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: will do

File: charters/PORTABILITY/charter.md (hunk 8 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: I think am going to remove the concept, please check next push.

File: charters/PORTABILITY/charter.md (hunk 12 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: I added a in scope item to help clarify what backends the team could add.

File: charters/PORTABILITY/charter.md (hunk 13 above)
Comment (TimothyTickle) [reply charters/PORTABILITY/charter.md]: These should be a part of the documentation of the service. I would prefer not to list them here as this may change and feels like hardcoding implementation detail.

**Comment by brianraymor** [comment]
@ambrosejcarr and @mckinsel  (since you are listed as the Technical Lead) - would you review the update and approve/reject?

File: charters/PipelinePortability/charter.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
+# [Pipeline Portability Service](mailto:pipelines-team@data.humancellatlas.org)
+
+## Description
+
+Reproducibility of analysis by others is a cornerstone of science. To satisfy this need, HCA data processing pipelines should be able to be executed by others in other projects and infrastructure. The Portability Service confirmations that pipelines developed for the HCA DCP can be successfully executed outside of the HCA DCP. Confirmation that HCA data processing pipelines can execute internally in the DCP is not assessed by this service but such confirmation is a part of our standard development and testing practices.
```
Comment (mckinsel) [inline charters/PipelinePortability/charter.md]: confirmations->confirms ?

File: charters/PipelinePortability/charter.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+## Objectives
+
+To confirm there are no technical barriers to executing pipelines outside of the DCP infrastructure, the Portability Service provides the ability to confirm HCA DCP pipelines can execute in other infrastructure. These infrastructure will be able to be added by the implementation team or 3rd parties.
```
Comment (mckinsel) [inline charters/PipelinePortability/charter.md]: maybe "..., the Portability Service executes pipelines in a set of other environments and verifies that they run successfully. The Portability Service implementation team as well as third parties can submit environments to be included in the Portability Service's test set."

**Review by mckinsel — APPROVED** [review]
Left a couple minor comments, but I think this covers it.

**Review by kbergin — APPROVED** [review]

File: charters/PipelinePortability/charter.md (hunk 15 above)
Comment (TimothyTickle) [reply charters/PipelinePortability/charter.md]: Updated to your suggestion.

File: charters/PipelinePortability/charter.md (hunk 14 above)
Comment (TimothyTickle) [reply charters/PipelinePortability/charter.md]: Thanks :-)

File: charters/PipelinePortability/charter.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+Portability service will not assess the scientific accuracy of pipelines, and only report on successful execution.  
+
+## Out-of-scope
+
+* Educational training or user assistance on how to create pipelines  
+* Updating backends so they support pipeline execution  
+* Updating pipelines so they are portable to environments  
+* Establishing standard data sets for portability (they are provided per pipeline)  
+* Adding 3rd party backends
```
Comment (samanehsan) [inline charters/PipelinePortability/charter.md:36]: Can you clarify why this is out of scope? The objectives state "The Portability Service implementation team as well as third parties can submit environments to be included in the Portability Service's test set" and one of the milestones is that "More than one backend are integrated" so it sounds like implementing the backends would be in scope. Perhaps there is something about the relationship between the portability service team and the third parties that could be stated here?

Comment (TimothyTickle) [reply charters/PipelinePortability/charter.md:36]: Implementing backends is in scope but not for 3rd parties. We would implement the backends we specifically need for the project but enable others to add to the system (or "register" themselves if they were interested.

Comment (samanehsan) [reply charters/PipelinePortability/charter.md:36]: Thanks Tim! My understanding is that implementing Cromwell support in the portability service is in scope because it is used for pipeline execution in the HCA. In addition, implementing at least one other backend is also in scope to demonstrate portability. But in general, the service would enable the developers of other backends to "register" themselves but it is out of scope for the portability team to add every possible backend. Does that sound right?

**Comment by mckinsel** [comment]
Approved at the 2019-03-08 tech architecture meeting.
